core/partitioning/partitioning.cpp

```diff
@@ -0,0 +1,250 @@
+    for (auto& input : seg_block.raw_inputs()) {
+      input_values.insert(input);
+    }
+  }
+
+  for (auto& graph_output : g->outputs()) {
+    input_values.insert(graph_output);
+  }
+
+  for (auto& mini_graph_input : input_values) {
+    for (auto& seg_block : segmented_blocks) {
+      if (std::find(seg_block.raw_inputs().begin(), seg_block.raw_inputs().end(), mini_graph_input) ==
+              seg_block.raw_inputs().end() &&
+          seg_block.contain_raw_input(mini_graph_input)) {
+        seg_block.registerOutput(mini_graph_input);
+      }
+    }
+  }
+
+  return;
+}
+
+void construct_segments(
+    std::vector<torch::jit::Node*>& pytorch_nodes,
+    std::vector<torch::jit::Node*>& tensorrt_nodes,
+    std::vector<SegmentedBlock>& segmented_blocks,
+    size_t min_block_size) {
+  // construct segmented blocks according to min_block_size and consecutive nodes
+  if (!tensorrt_nodes.empty()) {
+    if (tensorrt_nodes.size() < min_block_size) {
+      pytorch_nodes.insert(pytorch_nodes.end(), tensorrt_nodes.begin(), tensorrt_nodes.end());
+    } else {
+      if (!pytorch_nodes.empty())
+        segmented_blocks.emplace_back(SegmentedBlock::kTorch, pytorch_nodes);
+      segmented_blocks.emplace_back(SegmentedBlock::kTensorRT, tensorrt_nodes);
+      pytorch_nodes.clear();
+    }
+    tensorrt_nodes.clear();
+  }
+}
+
+std::vector<SegmentedBlock> segment_graph(
+    std::shared_ptr<torch::jit::Graph> g,
+    std::vector<conversion::InputRange>& input_ranges,
+    const conversion::TorchFallback& fallback_info) {
+  auto min_block_size = fallback_info.min_block_size;
+  std::unordered_set<std::string> forced_fallback_operators(
+      fallback_info.forced_fallback_operators.begin(), fallback_info.forced_fallback_operators.end());
+  std::vector<SegmentedBlock> segmented_blocks;
+
+  auto nodes = g->block()->nodes();
+
+  // segment the nodes
+  std::vector<torch::jit::Node*> tensorrt_nodes, pytorch_nodes;
+
+  for (const auto n : nodes) {
+    if (n->kind() == torch::jit::prim::Constant)
+      continue;
+
+    std::string node_string(n->kind().toQualString());
+    if (conversion::OpSupported(n) && !forced_fallback_operators.count(node_string)) {
+      tensorrt_nodes.push_back(n);
+    } else {
+      construct_segments(pytorch_nodes, tensorrt_nodes, segmented_blocks, min_block_size);
+      pytorch_nodes.push_back(n);
+    }
+  }
+  construct_segments(pytorch_nodes, tensorrt_nodes, segmented_blocks, min_block_size);
+  if (!pytorch_nodes.empty()) {
+    segmented_blocks.emplace_back(SegmentedBlock::kTorch, pytorch_nodes);
+  }
+
+  // register input/output torch::jit::Value for segmetned graphs
+  registerSegmentsInputsOutputs(segmented_blocks, g);
+
+  // store the mapping from lowering graph torch::jit::Value => torch::jit::IValue that we get by running segments
+  std::unordered_map<torch::jit::Value*, torch::jit::IValue> ivalues_maps;
+
+  std::vector<torch::jit::IValue> random_inputs = generateRandomInputs(input_ranges);

```

Comment:
Should we move this `generateRandomInputs`  function outside of fallback code for more general usage? something like utils ?

Reply:
Yes, I think it's ok. BTW, I found that my getFunctionSchema function in partitioning.cpp is same with the GenerateGraphSchema function in compiler.cpp, I am not sure if we should delete both of these two and move them to utils?